### PR TITLE
Functions with unboxed parameters

### PR DESCRIPTION
**Pull request by Ekdohibs**
This is an extension of #1248, with support for unboxed parameters as well as unboxed returns. Note that, for now, parameters will be unboxed if and only if there is an `[@unboxed]` annotation on the function, in which case the return is unboxed as well, if possible. Once we get #1257, we can unbox them on a per-parameter basis.

### CODE REVIEW

**Comment by mshinwell** [comment]
Example that now only boxes for the return value, as expected:
```
type t = { a : float; b : float }

let[@inline never] f x = { a = x; b = x *. 2. }

let foo y =
  let t1 = f y in
  let t2 = f (y *. 2.) in
   t1.a +. t1.b +. t2.a +. t2.b
```

**Comment by Ekdohibs** [comment]
This pull request is now mostly ready, and is waiting on a few other pull requests:
- #1268 for the backend changes that will be merged separately
- #1257 for attributes on parameters if we want to be able to specify exactly which parameters should be unboxed (and have warnings when they are not)
- optionally, #872 for per-parameter modes on lambda to be able to unbox local parameters (see CR in latest commit)

**Comment by Ekdohibs** [comment]
Rebased, now that #1257 has been merged.
There might be a couple problems with locals, because in for example:
```ocaml
let[@unboxed] f x =
  (exclave_ (local_ (x, x)))
```
the allocation becomes a heap allocation since `f.region` is false at the typedtree and lambda level. I'm not sure whether we can fix or even detect this without some significant changes, however: do we leave the responsibility to not misuse this to the user, or can this even cause segfaults due to the misuse of locals?

**Comment by Ekdohibs** [comment]
It looks like we will need to change a few things: with the following code, we end up with a heap-allocated value pointed to a local-allocated one:
```
let[@unboxed] f x =
  (exclave_ (let y = local_ (x, x) in local_ (y, y)))
```

**Comment by stedolan** [comment]
> the allocation becomes a heap allocation since f.region is false

I don't understand this - why does `f.region = false` cause an allocation to become a heap allocation?

**Comment by lthls** [comment]
> > the allocation becomes a heap allocation since f.region is false
> 
> I don't understand this - why does `f.region = false` cause an allocation to become a heap allocation?

(Note: I believe that it's when `f.region = true` that we're creating local allocations, but the problem still exists)

@Ekdohibs is on holidays, but if I remember correctly the issue is that the transformation doesn't actually remove allocations: it generates new allocations and field accesses, in places where the rest of the compiler is likely going to remove them.
Example:
```ocaml
(* Source version *)
let f : 'a -> 'a * 'a  = fun x ->
  *body*
[@@unboxed]

(* After transformation *)
let f_unboxed x =
  match *body* with
  | a, b -> #(a, b) (* Unboxed result *)

let f x =
  let #(a, b) = f_unboxed x in
  (a, b) (* Re-boxing *)
[@@inline always]
```

The issue is that we don't know the mode of the return value of the function at the point where we make the transformation, so we have to guess whether the re-boxing should be a local allocation or a heap allocation. Currently we use the `region` flag for that (called `contains_no_escaping_local_allocs` in flambda2), but it does not accurately reflect the expected mode of the return value, and we haven't found a correct solution yet.

File: middle_end/flambda2/from_lambda/lambda_to_flambda.ml [hunk 1]
```diff
@@ -1464,7 +1464,88 @@ and cps_function env ~fid ~(recursive : Recursive.t) ?precomputed_free_idents
...
+      Some (Unboxed_number bn)
+    | Pvalue (Pboxedvectorval bv) ->
+      let bn : Flambda_kind.Boxable_number.t =
+        match bv with Pvec128 _ -> Naked_vec128
+      in
+      Some (Unboxed_number bn)
+    | Pvalue (Pgenval | Pintval | Pvariant _ | Parrayval _)
+    | Ptop | Pbottom | Punboxed_float | Punboxed_int _ | Punboxed_vector _
+    | Punboxed_product _ ->
+      (* CR ncourant: warn *)
```
Comment (mshinwell) [inline middle_end/flambda2/from_lambda/lambda_to_flambda.ml:1496]: This should be fixed presumably?

File: middle_end/flambda2/from_lambda/closure_conversion.ml [hunk 2]
```diff
@@ -1400,9 +1400,68 @@ let close_switch acc env ~condition_dbg scrutinee (sw : IR.switch) :
...
+  match k with
+  | Multiple_values kinds ->
+    Flambda_primitive.Variadic
+      ( Make_block (Values (Tag.Scannable.zero, kinds), Immutable, alloc_mode),
+        Simple.vars unboxed_variables )
+  | Unboxed_number bn ->
+    let unboxed_variable =
+      match unboxed_variables with
+      | [var] -> var
+      | [] | _ :: _ :: _ -> assert false
```
Comment (Gbury) [inline middle_end/flambda2/from_lambda/closure_conversion.ml:1453]: Maybe a `Misc.fatal_errorf` with an informative message would be better here ?

File: middle_end/flambda2/from_lambda/closure_conversion.ml [hunk 3]
```diff
@@ -1695,11 +1927,258 @@ let close_one_function acc ~code_id ~external_env ~by_function_slot decl
...
+          else meta)
+        else meta
+      in
+      (* The wrapper doesn't contain any set of closures *)
+      let slot_offsets = Slot_offsets.empty in
+      ( wrapper_code,
+        Function_slot.Map.add main_function_slot main_approx by_function_slot,
+        (main_function_slot, main_code_id) :: function_code_ids,
+        Acc.add_code ~code_id:main_code_id ~code:main_code ~slot_offsets acc )
+  in
```
Comment (Gbury) [inline middle_end/flambda2/from_lambda/closure_conversion.ml:2181]: Could at least some of this code be split into separate functions ? This diff alone adds ~250 lines of uninterupted code to the body of a single function, and another diff above adds 200 lines to the same function. This makes the resulting code painful to read and understand, at least for me.

It is coherent with the rest of this module so I won't fight too hard on this, but if some parts of the code could be split into its own function, I think that would help.

**Review by Gbury — CHANGES_REQUESTED** [review]
The code looks good and works from the examples I tried, but there are a few places where I left some comments. Also, I can't answer the CR about locals and regions, maybe @lthls will know better.

**Comment by lthls** [comment]
(A small note to remember to check that attributes like `[@zero_alloc]` are attached to the correct function)

File: middle_end/flambda2/from_lambda/closure_conversion.ml [hunk 4]
```diff
@@ -1400,9 +1400,68 @@ let close_switch acc env ~condition_dbg scrutinee (sw : IR.switch) :
...
+          Flambda_kind.With_subkind.naked_float ))
+
+let unboxing_primitive (k : Function_decl.unboxing_kind) boxed_variable i =
+  match k with
+  | Multiple_values kinds ->
+    let block_access_kind : P.Block_access_kind.t =
+      Values
+        { tag = Known Tag.Scannable.zero;
+          size = Known (Targetint_31_63.of_int (List.length kinds));
+          field_kind = Any_value
```
[the remark this answers is not in the document]
Comment (Ekdohibs) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:1421]: This is a `Block_access_field_kind.t`, which is either `Immediate` or `Any_value` and currently always `Any_value` for block loads, so I'm not sure it's helpful to be more precise. The unboxed variable itself has a more precise kind, in the above function.

File: middle_end/flambda2/from_lambda/closure_conversion.ml (hunk 3 above)
Comment (Ekdohibs) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:2185]: I just tried to do that, I'm not sure if it's better or not, please look!

Comment (Gbury) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:2185]: That indeed helps in reading the diff, thanks !

File: middle_end/flambda2/from_lambda/closure_conversion.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Gbury) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:1425]: Seems reasonable, ^^

**Review by Gbury — APPROVED** [review]

**Pull request merged**